cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,300 @@
+    return avroFile;
+  }
+
+  private AvroFile createAvroFile(LogPathIdentifier identifier, long timestamp) throws IOException {
+    Location location = loggingLocationFactory.getLocation(identifier, timestamp);
+    LOG.info("Creating Avro file {}", location);
+    AvroFile avroFile = new AvroFile(location);
+    try {
+      avroFile.open();
+    } catch (IOException e) {
+      closeAndDelete(avroFile);
+      throw e;
+    }
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location);
+    } catch (Throwable e) {
+      closeAndDelete(avroFile);
+      throw new IOException(e);
+    }
+
+    fileMap.put(identifier, avroFile);
+    return avroFile;
+  }
+
+  private AvroFile rotateFile(AvroFile avroFile, LogPathIdentifier identifier, long timestamp) throws IOException {
+    long currentTs = System.currentTimeMillis();
+    long timeSinceFileCreate = currentTs - avroFile.getCreateTime();
+    if (timeSinceFileCreate > maxLifetimeMillis) {
+      avroFile.close();
+    }
+    if (!avroFile.isOpen()) {
+      LOG.info("Rotating a closed file {}", avroFile.getLocation());
+      return createAvroFile(identifier, timestamp);
+    }
+    return avroFile;
+  }
+
+  /**
+   * closes all the open avro files in the map
+   */
+  public void close() {
+    // close all the files in fileMap
+    // clear the map
+    for (AvroFile file : fileMap.values()) {
+      try {
+        file.flush();
+        file.close();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing and closing contents to file {}", file.getLocation(), e);
+      }
+    }
+    fileMap.clear();
+  }
+
+  private void closeAndDelete(AvroFile avroFile) {
+    try {
+      try {
+        avroFile.close();
+      } finally {
+        if (avroFile.getLocation().exists()) {
+          avroFile.getLocation().delete();
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while closing and deleting file {}", avroFile.getLocation(), e);
+    }
+  }
+
+  /**
+   * flushes the contents of all the open avro files
+   * @throws IOException
+   */
+  public void flush() throws IOException {
+    // perform flush on all the files in the fileMap
+    for (AvroFile file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation());
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Represents an Avro file.
+   *
+   * Since there is no way to check the state of the underlying file on an exception,
+   * all methods of this class assume that the file state is bad on any exception and close the file.
+   */
+  public class AvroFile implements Closeable {
+    private final Location location;
+    private FSDataOutputStream outputStream;
+    private DataFileWriter<GenericRecord> dataFileWriter;
+    private long createTime;
+    private boolean isOpen = false;
```

Comment:
Shouldn't have this variable. You should use the same contract as `OutputStream`, that the constructor would do the opening and will throw `IOException` if it fails to open the stream.